From 797771535266f29b60e7da660df8ed6460826377 Mon Sep 17 00:00:00 2001
From: Yaroslav Halchenko <debian@onerussian.com>
Date: Sun, 10 Jun 2018 10:58:36 -0400
Subject: [PATCH 1/2] BF: Centralize getting information about console width
 and set to 80 if 0

Also set terminal size to 80 if reported 0, and at least 20

- terminal size could be reported 0, thus causing the breakage due to -5 below
  See e.g. https://circleci.com/gh/conda-forge/datalad-feedstock/6
- we had multiple places were we used different ways to figure out the width

Now it should all use get_console_width defined in .ui.utils
---
 datalad/cmdline/main.py            |  5 ++---
 datalad/cmdline/tests/test_main.py |  5 ++---
 datalad/support/annexrepo.py       |  3 +--
 datalad/ui/utils.py                | 17 +++++++++++++++++
 4 files changed, 22 insertions(+), 8 deletions(-)

diff --git a/datalad/cmdline/main.py b/datalad/cmdline/main.py
index 11944827..4bb86305 100644
--- a/datalad/cmdline/main.py
+++ b/datalad/cmdline/main.py
@@ -7,6 +7,7 @@
 #
 # ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
 """"""
+from datalad.ui.utils import get_console_width
 
 __docformat__ = 'restructuredtext'
 
@@ -18,7 +19,6 @@ lgr.log(5, "Importing cmdline.main")
 import argparse
 import sys
 import textwrap
-import shutil
 from importlib import import_module
 import os
 
@@ -394,8 +394,7 @@ def get_description_with_cmd_summary(grp_short_descriptions, interface_groups,
     from ..interface.base import dedent_docstring
     lgr.debug("Generating detailed description for the parser")
     cmd_summary = []
-    console_width = shutil.get_terminal_size()[0] \
-        if hasattr(shutil, 'get_terminal_size') else 80
+    console_width = get_console_width()
     for i, grp in enumerate(
             sorted(interface_groups, key=lambda x: x[1])):
         grp_descr = grp[1]
diff --git a/datalad/cmdline/tests/test_main.py b/datalad/cmdline/tests/test_main.py
index 5fe27367..232ac7bd 100644
--- a/datalad/cmdline/tests/test_main.py
+++ b/datalad/cmdline/tests/test_main.py
@@ -22,6 +22,7 @@ import datalad
 from ..main import main, fail_with_short_help
 from datalad import __version__
 from datalad.cmd import Runner
+from datalad.ui.utils import get_console_width
 from datalad.api import create
 from datalad.utils import chpwd
 from datalad.tests.utils import with_tempfile
@@ -103,9 +104,7 @@ def test_help_np():
     # none of the lines must be longer than 80 chars
     # TODO: decide on   create-sibling and possibly
     # rewrite-urls
-    import shutil
-    accepted_width = shutil.get_terminal_size()[0] \
-        if hasattr(shutil, 'get_terminal_size') else 80
+    accepted_width = get_console_width()
 
     long_lines = ["%d %s" % (len(l), l) for l in stdout.split('\n')
                   if len(l) > accepted_width and
diff --git a/datalad/support/annexrepo.py b/datalad/support/annexrepo.py
index 300ea551..d004781e 100644
--- a/datalad/support/annexrepo.py
+++ b/datalad/support/annexrepo.py
@@ -3645,8 +3645,7 @@ class ProcessAnnexProgressIndicators(object):
                         j.get('percent-progress', '').rstrip('%')
                     ) or \
                     0
-            w, h = ui_utils.get_terminal_size()
-            w = w or 80  # default to 80
+            w = ui_utils.get_console_width()
             title = str(download_item)
             pbar_right = 50
             title_len = w - pbar_right - 4  # (4 for reserve)
diff --git a/datalad/ui/utils.py b/datalad/ui/utils.py
index e143da26..6b3e3c37 100644
--- a/datalad/ui/utils.py
+++ b/datalad/ui/utils.py
@@ -48,3 +48,20 @@ def get_terminal_size():
             return w, h
         except:
             return None, None
+
+
+def get_console_width(default_min=20):
+    """Return console width to use
+
+    In some cases shutil reports it to be 0, so we cannot rely on it
+    alone
+    """
+    console_width = get_terminal_size()[0] or 0
+    # it might still be 0, e.g. in conda builds for 0.10.0
+    if console_width <= 0:
+        console_width = 80
+    elif console_width <= 20:
+        # or some other too small to be real number,
+        # to prevent crashes below guarantee that it is at least 20
+        console_width = 20
+    return console_width
-- 
2.17.0

